File: mobile/android/app/src/main/java/com/nestorymobile/NestoryWidgetModule.kt
package com.nestorymobile

import android.content.Context
import androidx.glance.appwidget.updateAll
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

// Bridges dashboard summary data from JS into the shared storage the
// Glance home screen widget (NestoryWidget.kt) reads from. The widget has
// no network/auth access of its own — it only ever shows the last snapshot
// the app wrote, refreshed whenever the dashboard loads successfully.
class NestoryWidgetModule(reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext) {

  override fun getName() = "NestoryWidget"

  @ReactMethod
  fun writeSnapshot(json: String) {
    val prefs =
      reactApplicationContext.getSharedPreferences(WIDGET_PREFS, Context.MODE_PRIVATE)
    prefs.edit().putString(WIDGET_DATA_KEY, json).apply()

    CoroutineScope(Dispatchers.Main).launch {
      NestoryWidget().updateAll(reactApplicationContext)
    }
  }

  companion object {
    const val WIDGET_PREFS = "nestory_widget_prefs"
    const val WIDGET_DATA_KEY = "widget_data"
  }
}

File: mobile/ios/NestoryMobile/NestoryWidgetModule.swift
import Foundation
import WidgetKit

// Bridges dashboard summary data from JS into the shared App Group storage
// the WidgetKit extension (NestoryWidgetExtension/NestoryWidgetBundle.swift)
// reads from. The widget has no network/auth access of its own — it only
// ever shows the last snapshot the app wrote. WidgetCenter is Swift-only
// (no Objective-C header), which is why this module is Swift rather than
// the Objective-C most of this app's other native modules use.
@objc(NestoryWidgetModule)
class NestoryWidgetModule: NSObject {
  private let appGroupId = "group.com.nestoryhomekit.app"
  private let widgetDataKey = "widget_data"

  @objc
  func writeSnapshot(_ json: String) {
    let defaults = UserDefaults(suiteName: appGroupId)
    defaults?.set(json, forKey: widgetDataKey)

    if #available(iOS 14.0, *) {
      WidgetCenter.shared.reloadAllTimelines()
    }
  }

  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }
}
